lib/db_utils.go

```diff
@@ -283,21 +274,355 @@ var (
+			return errors.Wrapf(err, "DBDeleteWithTxn: Problem preparing ancestral record")
+		}
+		// Now delete the past record from the cache.
+		snap.DatabaseCache.Delete(keyString)
+		// We have to remove the previous value from the state checksum.
+		// Because checksum is commutative, we can safely remove the past value here.
+		snap.RemoveChecksumBytes(EncodeKeyValue(key, ancestralValue))
+	}
+	return nil
+}
+
+// DBIteratePrefixKeys fetches a chunk of records from the provided db at a provided prefix,
+// and beginning with the provided startKey. The chunk will have a total size of at least targetBytes.
+// If the startKey is a valid key in the db, it will be the first entry in the returned dbEntries.
+// If we have exhausted all entries for a prefix then _isChunkFull will be set as false, and true otherwise,
+// when there are more entries in the db at the prefix.
+func DBIteratePrefixKeys(db *badger.DB, prefix []byte, startKey []byte, targetBytes uint32) (
+	_dbEntries []*DBEntry, _isChunkFull bool, _err error) {
+	var dbEntries []*DBEntry
+	var totalBytes int
+	var isChunkFull bool
+
+	err := db.View(func(txn *badger.Txn) error {
+		opts := badger.DefaultIteratorOptions
+
+		// Iterate over the prefix as long as there are valid keys in the DB.
+		it := txn.NewIterator(opts)
+		defer it.Close()
+		for it.Seek(startKey); it.ValidForPrefix(prefix) && !isChunkFull; it.Next() {
+			item := it.Item()
+			key := item.Key()
+			// Add the key, value pair to our dbEntries list.
+			err := item.Value(func(value []byte) error {
+				dbEntries = append(dbEntries, KeyValueToDBEntry(key, value))
+				// If total amount of bytes in the dbEntries exceeds the target bytes size, we set the chunk as full.
+				totalBytes += len(key) + len(value)
+				if totalBytes > int(targetBytes) {
+					isChunkFull = true
+				}
+				return nil
+			})
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		// Return false for _isChunkFull to indicate that we shouldn't query this prefix again because
+		// something is wrong.
+		return nil, false, err
+	}
+	return dbEntries, isChunkFull, nil
+}
+
+// DBDeleteAllStateRecords is an auxiliary function that is used to clean up the state
+// before starting hyper sync.
+func DBDeleteAllStateRecords(db *badger.DB) error {
+	// Iterate over all state prefixes.
+	for _, prefix := range StatePrefixes.StatePrefixesList {
+		startKey := prefix
+		fetchingPrefix := true
+
+		// We will delete all records for a prefix step by step. We do this in chunks of 8MB,
+		// to make sure we don't overload badger DB with the size of our queries. Whenever a
+		// chunk is not full, that is isChunkFull = false, it means that we've exhausted all
+		// entries for a prefix.
+		for ;fetchingPrefix; {
+			// Fetch a chunk of data from the DB.
+			dbEntries, isChunkFull, err := DBIteratePrefixKeys(db, prefix, startKey, 8 << 20)
+			fetchingPrefix = isChunkFull
```

Comment:
Ah yeah, this one is slightly tricky. So the DBIteratePrefixKeys routine will fill fetch entries until it reaches the provided byte size, in this case `8 << 20` (replaced by `SnapshotBatchSize`). So if there are more entries in the DB, then the fetched chunk will be full, or, `isChunkFull = true`. Otherwise, if we've exhausted all entries, we will not reach the byte size, i.e. `isChunkFull = false` and we can exit the loop.